docs/cookbook/dynamic-configuration/01_Basics/01_Structure.md

```diff
@@ -0,0 +1,65 @@
+# 01 Structure
+
+Dynamic Configuration uses a series of files to load templates. This example will show how they all combine together. Running the below command will combine all the templates into the final.yml.
+
+`docker run -v ${PWD}/:/etc/grafana grafana/agentctl:latest template-parse /etc/grafana/02_config.yml`
+
+## Dynamic Configuration
+
+[config.yml](./01_config.yml)
+
+```yaml
+template_paths:
+  - "file:///etc/grafana/01_assets"
+```
+
+Tells the Grafana Agent where to load files from. It is important to note that dynamic configuration does NOT traverse directories. It will look at the directory specified only, if you need more directories then add them to the `template_paths` array.
+
+## Agent
+
+Dynamic Configuration will find the first file matching pattern `agent-*.yml` and load that as the base. You can only have
```

Comment:
Is `-config.file` no longer required with `-dynamic-config`?  I'm not a fan of the naming prefixes and to me it makes more sense to pass the main config file to `-config.file` and have a `-config.directory` flag to import additional *.yaml files. (or configure this directory in the main config).

Reply:
If you are using dynamic configuration then `-config-file` points to the config file that handles the dynamic configuration, basically a list of folders to scan and any datasources you want to define. We could have `-config.directory` but you would either need to allow an array of them and also additional flags to specify datasources. 

You also do not need to specify an `agent-*.yml` at all, only import the parts you need.

Reply:
Oh I see.  IMO overriding the `-config.file` flag seems strange -- could a dynamic config block be including in the main agent config (what we're loading as `agent-*.yml` according to the proposed design)

Reply:
The dynamic config block tells the system where to find the `agent-*.yml` if we didn't want to reuse `-config.file` then a new flag could be added. I resued it for simplicity but if it makes it clearer then that's worth it. The dynamic config tells the system where to load the `agent-*.yml` so it needs to be added first. 

Example `/etc/grafana/dynamic-config.yml`
```
template_paths:
  - "file:///etc/grafana/config"
```

The directory `/etc/grafana/config` has one file `agent-1.yml`. 

Running `./agent -config.file=/etc/grafana/dynamic-config.yml -enable-features=dynamic-config,integrations-next` will 

1. Load the configuration at `/etc/grafana/dynamic-config.yml`
2. Search the directory at `/etc/grafana/config`
3. Find the `agent-1.yml`
4. Unmarshal the config and start the agent

Reply:
My thinking is that it makes more sense to do the following:

1. Load the `agent.yaml` config by passing it to `-config.file`.  This would contain the contents of `agent-*.yml` as well as the contents of `dynamic-config.yaml`
2. If the `dynamic-config` feature is enabled, parse the template paths.

This seems more "conventional" to me. It would mean we don't need a separate config when using dynamic config, and we don't have to discover the base agent config file.  I may be missing something here, but does this seem sane to you?

Reply:
My thought is you will want to have a base agent config file that itself will be templated/re-used. The benefit of having them separate is allowing you to configure a fleet of agent instances.  

If you have a dozen agent instances running and want to have a base agent that has organization defaults (endpoints, username, password, agent metrics enabled), then you can have a folder with just your `agent-1.yml` and in the dynamic configuration, template_paths you have the directory where your base `agent-1.yml` resides then the individual instances would have different directories they load the specific details of what they want to scrape/observe. The dynamic-configuration template_paths directories could even be templated with environment vars. 

```
template_paths:
  - "s3://agent_configs/base" # contains agent-1.yml
  - "s3://agent_configs/{{agent_purpose}} # contains specific overrides for this agents purpose
  - "s3://agent_configs/{{hostname}} # contains specific overrides for this host
```

If you needed to globally update a value or setting you could do it in the base and when the reload timer (as of yet undiscussed) triggered it would propagate. 

This might be a use case that is not useful or I am overcomplicating it =)

Reply:
Why can't the base still be templated or re-used if you combine the dynamic config?  I think you'd still be able to accomplish what you've described here (e.g. env vars in the template_paths section of the agent.yml) if you combine the dynamic config with the base config. It does sound a bit overcomplicated to me 😅